File: win/DoorbellApp/BootConfig.cs
using System.Collections.Generic;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Web.Script.Serialization;

namespace DoorbellApp
{
    public sealed class BootConfig
    {
        public string RawJson = "{}";
        public string Name = "doorbell";
        public string Role = "door_station";
        public string Door = "";
        public string UiLang = "ja";
        public string PskRef = "";
        public bool Kiosk = true;
        public int HttpPort = 47180;
        public string FilePath;
        public bool SetupRequired;
        public string SuggestedDoor;

        private const string DefaultJson =
            "{ \"name\": \"doorbell-win\", \"role\": \"door_station\", \"door\": \"\", " +
            "\"listen_port\": 47172, \"http_port\": 47180, \"ui_lang\": \"ja\", " +
            "\"kiosk\": false, \"setup_complete\": false }";

        public static BootConfig Load(string path)
        {
            var c = new BootConfig();
            c.FilePath = path;
            string primary = ReadValidJson(path);
            string backup = ReadValidJson(path + ".bak");
            c.RawJson = primary ?? backup ?? DefaultJsonWithSuggestedDoor();
            if (primary == null) WriteAtomic(path, c.RawJson);
            // A readable legacy profile is not proof that an operator confirmed
            // its local identity. Require the explicit marker after an upgrade.
            bool setupComplete = false;
            try
            {
                var d = new JavaScriptSerializer().Deserialize<Dictionary<string, object>>(c.RawJson);
                if (d != null)
                {
                    if (d.TryGetValue("name", out var n) && n != null) c.Name = n.ToString();
                    if (d.TryGetValue("role", out var ro) && ro != null) c.Role = ro.ToString();
                    if (d.TryGetValue("door", out var dr) && dr != null) c.Door = dr.ToString();
                    if (d.TryGetValue("ui_lang", out var l) && l != null) c.UiLang = l.ToString();
                    if (d.TryGetValue("psk_ref", out var pr) && pr != null) c.PskRef = pr.ToString();
                    if (d.TryGetValue("kiosk", out var k) && k is bool kb) c.Kiosk = kb;
                    if (d.TryGetValue("setup_complete", out var setup) && setup is bool sb)
                        setupComplete = sb;
                    if (d.TryGetValue("http_port", out var hp) && hp != null)
                    {
                        int p;
                        if (int.TryParse(hp.ToString(), out p) && p > 0) c.HttpPort = p;
                    }
                }
            }
            catch { }
            c.SuggestedDoor = ValidDoor(c.Door) ? c.Door : SuggestedDoorId();
            c.SetupRequired = !setupComplete || !ValidRole(c.Role) ||
                              (c.Role == "door_station" && !ValidDoor(c.Door));
            return c;
        }

        public static bool ValidRole(string value)
        {
            return value == "door_station" || value == "indoor_panel";
        }

        public static bool ValidDoor(string value)
        {
            if (string.IsNullOrEmpty(value) || value.Length > 64) return false;
            char first = value[0];
            if (!((first >= 'a' && first <= 'z') || (first >= 'A' && first <= 'Z') ||
                  (first >= '0' && first <= '9'))) return false;
            for (int i = 0; i < value.Length; ++i)
            {
                char ch = value[i];
                if (!((ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') ||
                      (ch >= '0' && ch <= '9') || ch == '_' || ch == '-')) return false;
            }
            return true;
        }

        private static string SuggestedDoorId()
        {
            byte[] bytes = new byte[4];
            using (var random = RandomNumberGenerator.Create()) random.GetBytes(bytes);
            var text = new StringBuilder("door-");
            foreach (byte value in bytes) text.Append(value.ToString("x2"));
            return text.ToString();
        }

        private static string DefaultJsonWithSuggestedDoor()
        {
            try
            {
                var json = new JavaScriptSerializer();
                var data = json.Deserialize<Dictionary<string, object>>(DefaultJson) ??
                           new Dictionary<string, object>();
                data["door"] = SuggestedDoorId();
                return json.Serialize(data);
            }
            catch { return DefaultJson; }
        }

        public static BootConfig PersistSetup(string path, string name, string role, string door)
        {
            role = (role ?? "").Trim();
            door = (door ?? "").Trim();
            if (!ValidRole(role) || (role == "door_station" && !ValidDoor(door))) return null;
            name = (name ?? "").Trim();
            if (name.Length > 64) name = name.Substring(0, 64);
            if (name.Length == 0) name = "doorbell";
            string current = ReadValidJson(path) ?? ReadValidJson(path + ".bak") ?? DefaultJson;
            try
            {
                var json = new JavaScriptSerializer();
                var data = json.Deserialize<Dictionary<string, object>>(current) ??
                           new Dictionary<string, object>();
                data["name"] = name;
                data["role"] = role;
                data["door"] = role == "door_station" ? door : "";
                data["setup_complete"] = true;
                return WriteAtomic(path, json.Serialize(data)) ? Load(path) : null;
            }
            catch { return null; }
        }

        /// <summary>Pairing PSK/seeds are crash-safely persisted for the next process start.</summary>
        public static string PersistPairingReference(string path, IEnumerable<string> seeds)
        {
            string current = ReadValidJson(path) ?? ReadValidJson(path + ".bak") ?? DefaultJson;
            try
            {
                var json = new JavaScriptSerializer();
                var data = json.Deserialize<Dictionary<string, object>>(current) ??
                           new Dictionary<string, object>();
                data.Remove("psk_hex");
                data["psk_ref"] = "secret:mesh.psk";
                var merged = new List<string>();
                object existing;
                if (data.TryGetValue("seed_peers", out existing) &&
                    existing is System.Collections.IEnumerable && !(existing is string))
                    foreach (object value in (System.Collections.IEnumerable)existing)
                        AddUnique(merged, value == null ? null : value.ToString());
                if (seeds != null) foreach (string seed in seeds) AddUnique(merged, seed);
                if (merged.Count != 0) data["seed_peers"] = merged;
                string output = json.Serialize(data);
                return WritePairingGenerations(path, output) ? output : null;
            }
            catch { return null; }
        }

        /// <summary>Drops the pairing reference and seeds after unpair/revoke.</summary>
        public static bool ClearPairingReference(string path)
        {
            string current = ReadValidJson(path) ?? ReadValidJson(path + ".bak") ?? DefaultJson;
            try
            {
                var json = new JavaScriptSerializer();
                var data = json.Deserialize<Dictionary<string, object>>(current) ??
                           new Dictionary<string, object>();
                data.Remove("psk_hex");
                data.Remove("psk_ref");
                data.Remove("seed_peers");
                return WritePairingGenerations(path, json.Serialize(data));
            }
            catch { return false; }
        }

        /// <summary>
        /// Revoke and device-side "leave the Cluster" are a factory reset of this device's
        /// identity (spec 5.4): the pairing reference, the discovery seeds, and the operator's own
        /// name/role/door confirmation all go, so the next start lands in first-run setup.
        /// </summary>
        public static bool ResetToFactory(string path)
        {
            try
            {
                var json = new JavaScriptSerializer();
                var current = json.Deserialize<Dictionary<string, object>>(
                    ReadValidJson(path) ?? ReadValidJson(path + ".bak") ?? DefaultJson) ??
                    new Dictionary<string, object>();
                var data = new Dictionary<string, object>();
                // Only transport-level bootstrap survives; nothing that identifies this device or
                // its Cluster does.
                foreach (string keep in new[] { "listen_port", "http_port", "ui_lang", "kiosk" })
                {
                    object value;
                    if (current.TryGetValue(keep, out value) && value != null) data[keep] = value;
                }
                data["name"] = "doorbell";
                data["role"] = "";
                data["door"] = SuggestedDoorId();
                data["setup_complete"] = false;
                return WritePairingGenerations(path, json.Serialize(data));
            }
            catch { return false; }
        }

        private static void AddUnique(List<string> values, string value)
        {
            if (string.IsNullOrWhiteSpace(value)) return;
            value = value.Trim();
            if (!values.Contains(value)) values.Add(value);
        }

        private static string ReadValidJson(string path)
        {
            try
            {
                if (!File.Exists(path)) return null;
                string text = File.ReadAllText(path, Encoding.UTF8);
                var parsed = new JavaScriptSerializer().Deserialize<Dictionary<string, object>>(text);
                return parsed == null ? null : text;
            }
            catch { return null; }
        }

        private static bool WriteAtomic(string path, string value)
        {
            string temporary = path + ".tmp";
            try
            {
                string directory = Path.GetDirectoryName(path);
                if (!string.IsNullOrEmpty(directory)) Directory.CreateDirectory(directory);
                WriteGeneration(path, value);
                return true;
            }
            catch
            {
                try { File.Delete(temporary); } catch { }
                return false;
            }
        }

        private static bool WritePairingGenerations(string path, string value)
        {
            try
            {
                string directory = Path.GetDirectoryName(path);
                if (!string.IsNullOrEmpty(directory)) Directory.CreateDirectory(directory);
                // Both recoverable generations must contain the scrubbed secret reference.
                WriteGeneration(path + ".bak", value);
                WriteGeneration(path, value);
                return true;
            }
            catch
            {
                try { File.Delete(path + ".tmp"); } catch { }
                try { File.Delete(path + ".bak.tmp"); } catch { }
                return false;
            }
        }

        private static void WriteGeneration(string target, string value)
        {
            string temporary = target + ".tmp";
            byte[] bytes = Encoding.UTF8.GetBytes(value);
            using (var output = new FileStream(temporary, FileMode.Create, FileAccess.Write,
                                               FileShare.None, 4096, FileOptions.WriteThrough))
            {
                output.Write(bytes, 0, bytes.Length);
                output.Flush(true);
            }
            if (File.Exists(target)) File.Replace(temporary, target, null, true);
            else File.Move(temporary, target);
        }
    }
}

File: android/app/src/main/java/jp/keihan/doorbell/BootConfig.kt
// boot.json contains local startup settings; fleet configuration remains in Core's CRDT.
package jp.keihan.doorbell

import java.io.File
import java.io.FileOutputStream
import java.nio.charset.Charset
import java.security.SecureRandom
import org.json.JSONObject

class BootConfig private constructor(
    val rawJson: String,
    val name: String,
    val role: String,
    val door: String,
    val uiLang: String,
    val kiosk: Boolean,
    val kioskMode: String,
    val bootLaunch: Boolean,
    val rootHelper: Boolean,
    val watchdog: Boolean,
    val httpPort: Int,
    /** True until an operator has confirmed a valid local role/profile. */
    val setupRequired: Boolean,
    /** A valid, unique-enough suggestion for a door-station assignment. */
    val suggestedDoor: String,
) {
    companion object {
        private const val DEFAULT_JSON =
            "{ \"name\": \"doorbell-android\", \"role\": \"door_station\", \"door\": \"\", " +
            "\"listen_port\": 47172, \"http_port\": 47180, \"ui_lang\": \"ja\", \"kiosk\": false, " +
            "\"setup_complete\": false }"

        private val random = SecureRandom()

        private fun suggestedDoorId(): String {
            val bytes = ByteArray(4)
            random.nextBytes(bytes)
            return "door-" + bytes.joinToString("") { "%02x".format(it.toInt() and 0xff) }
        }

        fun validRole(value: String): Boolean =
            value == "door_station" || value == "indoor_panel"

        fun validDoor(value: String): Boolean =
            value.matches(Regex("[A-Za-z0-9][A-Za-z0-9_-]{0,63}"))

        fun load(file: File): BootConfig {
            val backup = File(file.parentFile, file.name + ".bak")
            val primary = readValidJson(file)
            val recovered = if (primary == null) readValidJson(backup) else null
            val raw = primary ?: recovered ?: JSONObject(DEFAULT_JSON).apply {
                put("door", suggestedDoorId())
            }.toString()
            if (primary == null) writeAtomic(file, raw, preserveCurrent = false)
            var name = "doorbell"; var role = "door_station"; var door = ""
            var lang = "ja"; var kiosk = false; var kioskMode = "auto"
            var bootLaunch = true; var rootHelper = false; var watchdog = true
            var httpPort = 47180
            // A readable legacy profile is not proof that an operator confirmed
            // its local identity.  Require the explicit marker so upgrades also
            // surface the setup UI once instead of silently accepting a blank or
            // accidentally inherited role.
            var setupComplete = false
            try {
                val d = JSONObject(raw)
                name = d.optString("name", name)
                role = d.optString("role", role)
                door = d.optString("door", door)
                lang = d.optString("ui_lang", lang)
                when (val value = d.opt("kiosk")) {
                    is JSONObject -> {
                        kiosk = value.optBoolean("enabled", true)
                        kioskMode = normalizeKioskMode(value.optString("mode", kioskMode))
                        bootLaunch = value.optBoolean("boot_launch", bootLaunch)
                        rootHelper = value.optBoolean("root_helper", rootHelper)
                        watchdog = value.optBoolean("watchdog", watchdog)
                    }
                    is Boolean -> kiosk = value
                }
                httpPort = d.optInt("http_port", httpPort)
                if (d.has("setup_complete")) setupComplete = d.optBoolean("setup_complete", false)
            } catch (_: Exception) { }
            val roleValid = validRole(role)
            val doorValid = validDoor(door)
            val suggestedDoor = if (doorValid) door else suggestedDoorId()
            val setupRequired = !setupComplete || !roleValid ||
                (role == "door_station" && !doorValid)
            return BootConfig(raw, name, role, door, lang, kiosk, kioskMode, bootLaunch,
                              rootHelper, watchdog, httpPort, setupRequired, suggestedDoor)
        }

        /**
         * Saves the locally-owned bootstrap identity without touching pairing secrets or seeds.
         * Door stations require a stable nonempty door ID; indoor panels deliberately do not.
         */
        fun persistSetup(file: File, name: String, role: String, door: String): BootConfig? {
            if (!validRole(role)) return null
            val normalizedName = name.trim().take(64).ifEmpty { "doorbell" }
            val normalizedDoor = door.trim()
            if (role == "door_station" && !validDoor(normalizedDoor)) return null
            val cur = readValidJson(file)
                ?: readValidJson(File(file.parentFile, file.name + ".bak"))
                ?: DEFAULT_JSON
            return try {
                val json = JSONObject(cur)
                json.put("name", normalizedName)
                json.put("role", role)
                json.put("door", if (role == "door_station") normalizedDoor else "")
                json.put("setup_complete", true)
                val text = json.toString()
                if (!writeAtomic(file, text, preserveCurrent = true)) null else load(file)
            } catch (_: Exception) { null }
        }

        /** Persist only the secure-store reference and seeds. Both boot generations are scrubbed. */
        fun persistPskReference(file: File, seeds: List<String>): String? {
            val cur = readValidJson(file)
                ?: readValidJson(File(file.parentFile, file.name + ".bak"))
                ?: DEFAULT_JSON
            return try {
                val d = JSONObject(cur)
                d.remove("psk_hex")
                d.put("psk_ref", "secret:mesh.psk")
                val merged = LinkedHashSet<String>()
                d.optJSONArray("seed_peers")?.let { arr ->
                    for (i in 0 until arr.length()) arr.optString(i).takeIf { it.isNotEmpty() }
                        ?.let { merged.add(it) }
                }
                for (s in seeds) if (s.isNotEmpty()) merged.add(s)
                if (merged.isNotEmpty()) d.put("seed_peers", org.json.JSONArray(merged.toList()))
                val js = d.toString()
                val backup = File(file.parentFile, file.name + ".bak")
                writeAndRename(backup, js)
                writeAndRename(file, js)
                js
            } catch (_: Exception) { null }
        }

        /**
         * Drop the cluster secret reference and seeds after core reported state "unpaired".
         * Returns the rewritten boot JSON, or null when nothing had to change or the write failed.
         */
        fun clearPskReference(file: File): String? {
            val cur = readValidJson(file)
                ?: readValidJson(File(file.parentFile, file.name + ".bak"))
                ?: return null
            return try {
                val d = JSONObject(cur)
                if (!d.has("psk_ref") && !d.has("psk_hex") && !d.has("seed_peers")) return null
                d.remove("psk_ref")
                d.remove("psk_hex")
                d.remove("seed_peers")
                val js = d.toString()
                writeAndRename(File(file.parentFile, file.name + ".bak"), js)
                writeAndRename(file, js)
                js
            } catch (_: Exception) { null }
        }

        fun hasSecureMeshReference(rawJson: String): Boolean = try {
            JSONObject(rawJson).optString("psk_ref") == "secret:mesh.psk"
        } catch (_: Exception) { false }

        private fun normalizeKioskMode(value: String): String = when (value) {
            "auto", "native", "root_helper", "home_only" -> value
            else -> "auto"
        }

        private fun readValidJson(file: File): String? {
            if (!file.isFile) return null
            return try {
                val text = file.readText(Charsets.UTF_8)
                JSONObject(text)
                text
            } catch (_: Exception) { null }
        }

        /** Write through a same-directory temporary and keep the last valid generation. */
        private fun writeAtomic(file: File, value: String, preserveCurrent: Boolean): Boolean {
            val parent = file.parentFile ?: return false
            if (!parent.exists() && !parent.mkdirs()) return false
            return try {
                if (preserveCurrent) {
                    readValidJson(file)?.let { current ->
                        writeAndRename(File(parent, file.name + ".bak"), current)
                    }
                }
                writeAndRename(file, value)
                true
            } catch (_: Exception) { false }
        }

        private fun writeAndRename(target: File, value: String) {
            val tmp = File(target.parentFile, target.name + ".tmp")
            FileOutputStream(tmp).use { out ->
                out.write(value.toByteArray(Charsets.UTF_8))
                out.flush()
                out.fd.sync()
            }
            if (!tmp.renameTo(target)) {
                tmp.delete()
                throw java.io.IOException("rename failed: ${target.name}")
            }
        }
    }
}
